compiler/rustc_hir_analysis/src/astconv/errors.rs

```diff
@@ -597,7 +597,28 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                 }
             }
         }
-        if !suggestions.is_empty() {
+        suggestions.sort_by_key(|&(span, _)| span);
+        // There are cases where one bound points to a span within another bound's span, like when
+        // you have code like the following (#115019), so we skip providing a suggestion in those
+        // cases to avoid having a malformed suggestion.
+        //
+        // pub struct Flatten<I> {
+        //     inner: <IntoIterator<Item: IntoIterator<Item: >>::IntoIterator as Item>::core,
+        //             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+        //             |                  ^^^^^^^^^^^^^^^^^^^^^
+        //             |                  |
+        //             |                  associated types `Item`, `IntoIter` must be specified
+        //             associated types `Item`, `IntoIter` must be specified
+        // }
+        let mut overlaps = false;
+        let mut sugg = suggestions.iter().peekable();
+        while let Some((span, _)) = sugg.next() {
+            let Some((next, _)) = sugg.peek() else {
+                break;
+            };
+            overlaps |= span.overlaps(*next);
+        }
```

Comment:
I think this loop can be simplified? Will need to be tweaked if the above assumption is not true, though.

```suggestion
        let overlaps = suggestions.windows(2).any(|pair| pair[0].0.overlaps(pair[1].0));
```